spark/src/main/java/com/netflix/iceberg/spark/data/SparkParquetWriters.java

```diff
@@ -0,0 +1,447 @@
+      List<Type> fields = struct.getFields();
+      List<ParquetValueWriter<?>> writers = Lists.newArrayListWithExpectedSize(fieldWriters.size());
+      List<DataType> sparkTypes = Lists.newArrayList();
+      for (int i = 0; i < fields.size(); i += 1) {
+        Type fieldType = struct.getType(i);
+        int fieldD = type.getMaxDefinitionLevel(path(fieldType.getName()));
+        writers.add(option(fieldType, fieldD, fieldWriters.get(i)));
+        sparkTypes.add(convert(schema.findType(fieldType.getId().intValue())));
+      }
+
+      return new InternalRowWriter(writers, sparkTypes);
+    }
+
+    @Override
+    public ParquetValueWriter<?> list(GroupType array, ParquetValueWriter<?> elementWriter) {
+      GroupType repeated = array.getFields().get(0).asGroupType();
+      String[] repeatedPath = currentPath();
+
+      int repeatedD = type.getMaxDefinitionLevel(repeatedPath);
+      int repeatedR = type.getMaxRepetitionLevel(repeatedPath);
+
+      org.apache.parquet.schema.Type elementType = repeated.getType(0);
+      int elementD = type.getMaxDefinitionLevel(path(elementType.getName()));
+
+      DataType elementSparkType = convert(schema.findType(elementType.getId().intValue()));
+
+      return new ArrayDataWriter<>(repeatedD, repeatedR,
+          option(elementType, elementD, elementWriter),
+          elementSparkType);
+    }
+
+    @Override
+    public ParquetValueWriter<?> map(GroupType map,
+                                     ParquetValueWriter<?> keyWriter,
+                                     ParquetValueWriter<?> valueWriter) {
+      GroupType repeatedKeyValue = map.getFields().get(0).asGroupType();
+      String[] repeatedPath = currentPath();
+
+      int repeatedD = type.getMaxDefinitionLevel(repeatedPath);
+      int repeatedR = type.getMaxRepetitionLevel(repeatedPath);
+
+      org.apache.parquet.schema.Type keyType = repeatedKeyValue.getType(0);
+      int keyD = type.getMaxDefinitionLevel(path(keyType.getName()));
+      DataType keySparkType = convert(schema.findType(keyType.getId().intValue()));
+      org.apache.parquet.schema.Type valueType = repeatedKeyValue.getType(1);
+      int valueD = type.getMaxDefinitionLevel(path(valueType.getName()));
+      DataType valueSparkType = convert(schema.findType(valueType.getId().intValue()));
+
+      return new MapDataWriter<>(repeatedD, repeatedR,
+          option(keyType, keyD, keyWriter), option(valueType, valueD, valueWriter),
+          keySparkType, valueSparkType);
+    }
+
+    @Override
+    public ParquetValueWriter<?> primitive(PrimitiveType primitive) {
+      ColumnDescriptor desc = type.getColumnDescription(currentPath());
+
+      if (primitive.getOriginalType() != null) {
+        switch (primitive.getOriginalType()) {
+          case ENUM:
+          case JSON:
+          case UTF8:
+            return utf8Strings(desc);
+          case DATE:
+          case INT_8:
+          case INT_16:
+          case INT_32:
+          case INT_64:
```

Comment:
I realize that INT96 is deprecated as per https://issues.apache.org/jira/browse/PARQUET-323 and don't want to encourage people using it in Iceberg but if people do want to re-write data with INT96 rows into iceberg (as INT64)  would we run into this code? If so should we not handle it by truncating INT96 to INT64?

Reply:
No, by the time this code gets INT96 values, they will be converted to binary, fixed, or a long. That's the engine's job because Iceberg won't be used to read INT96 values from files. So another read implementation produces values, then the engine hands them to Iceberg using a known type. No need to handle it here.